src/ast/semantic_analyser.cpp

```diff
@@ -170,6 +170,26 @@ void SemanticAnalyser::visit(Builtin &builtin)
       buf << arch::name() << " doesn't support " << builtin.ident;
     builtin.type = SizedType(Type::integer, 8);
   }
+  else if (!builtin.ident.compare(0, 4, "sarg") && builtin.ident.size() == 5 &&
+      builtin.ident.at(4) >= '0' && builtin.ident.at(4) <= '9') {
+    for (auto &attach_point : *probe_->attach_points)
+    {
+      ProbeType type = probetype(attach_point->provider);
+      if (type != ProbeType::kprobe && type != ProbeType::uprobe)
+        err_ << "The " << builtin.ident << " builtin can only be used with "
+             << "'kprobes' and 'uprobes' probes" << std::endl;
+
+      if (is_final_pass() &&
+          (attach_point->address != 0 || attach_point->func_offset != 0)) {
+        std::stringstream warn_buf;
+        warn_buf << "Using an address offset with the sargX built-in can lead "
+                 << "to unexpected behavior. Consider using a probe without an "
+                 << "offset.";
```

Comment:
I think just having the first line is enough. A probe without offset isn't really an alternative you can use if you need offsets. The alternative you do have is to store `sargs` in a map on function entry and use that in the offset probe. But I think we're  better off documenting that somewhat instead of trying to put it all in the warning message.